File: Medium/725. Split Linked List in Parts/725. Split Linked List in Parts.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution(object):
    def splitListToParts(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: List[ListNode]
        """
        # Calculate the length of the linked list
        length = 0
        current = head
        while current:
            length += 1
            current = current.next

        # Calculate the size of each part and the number of extra nodes
        part_size = length // k
        extra_nodes = length % k

        result = []
        current = head

        for i in range(k):
            result.append(current)
            part_length = part_size + 1 if i < extra_nodes else part_size

            # Move current to the end of this part
            for j in range(part_length - 1):
                current = current.next

            if current:
                # Save the next node, break the link, and move to the next part
                next_node = current.next
                current.next = None
                current = next_node

        return result

File: Medium/725. Split Linked List in Parts/725. Split Linked List in Parts.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
public class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {
        int length = 0;
        ListNode current = head;
        
        // Calculate the length of the linked list
        while (current != null) {
            length++;
            current = current.next;
        }
        
        int partSize = length / k;
        int extraNodes = length % k;
        
        ListNode[] result = new ListNode[k];
        current = head;
        
        for (int i = 0; i < k; i++) {
            result[i] = current;
            int partLength = partSize + (extraNodes > 0 ? 1 : 0);
            
            // Move current to the end of this part
            for (int j = 1; j < partLength && current != null; j++) {
                current = current.next;
            }
            
            // If there are extra nodes, decrement extraNodes
            if (extraNodes > 0) {
                extraNodes--;
            }
            
            // Save the next node, break the link, and move to the next part
            if (current != null) {
                ListNode nextNode = current.next;
                current.next = null;
                current = nextNode;
            }
        }
        
        return result;
    }
}